src/main/java/jenkins/jmh/JmhBenchmarkState.java

```diff
@@ -0,0 +1,163 @@
+package jenkins.jmh;
+
+import hudson.WebAppMain;
+import hudson.model.DownloadService;
+import hudson.model.Hudson;
+import hudson.model.JDK;
+import hudson.model.RootAction;
+import hudson.model.UpdateSite;
+import hudson.security.ACL;
+import hudson.util.PersistedList;
+import io.vavr.Tuple2;
+import jenkins.model.Jenkins;
+import jenkins.model.JenkinsLocationConfiguration;
+import org.eclipse.jetty.server.Server;
+import org.jvnet.hudson.test.JavaNetReverseProxy;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.TemporaryDirectoryAllocator;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+
+import javax.annotation.CheckForNull;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletContextEvent;
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@State(Scope.Benchmark)
+public abstract class JmhBenchmarkState implements RootAction {
+    private static final Logger LOGGER = Logger.getLogger(JmhBenchmarkState.class.getName());
+    private static final String contextPath = "/jenkins";
+
+    private final TemporaryDirectoryAllocator temporaryDirectoryAllocator = new TemporaryDirectoryAllocator();
+
+    private AtomicInteger localPort = new AtomicInteger();
+    private Jenkins jenkins = null;
+    private Server server = null;
+
+    // Run the setup for each individual fork of the JVM
+    @Setup(org.openjdk.jmh.annotations.Level.Trial)
```

Comment:
would be nice to do some refactoring to avoid code duplication with JenkinsRule

Reply:
I did do some refactoring. If you approve what I've done to `JenkinsRule`, I'll continue with the rest.

Reply:
Looks good so far